src/Microsoft.DotNet.Build.CloudTestTasks/AssemblyResolver.cs

```diff
@@ -0,0 +1,44 @@
+﻿using System;
+using System.IO;
+using System.Reflection;
+using System.Threading;
+
+namespace Microsoft.DotNet.Build.CloudTestTasks
+{
+    /// <summary>
+    /// Used to enable custom assembly resolution.
+    /// </summary>
+    class AssemblyResolver
+    {
+        static int s_enabled;
+
+        /// <summary>
+        /// Call to enable the assembly resolver for the current AppDomain.
+        /// </summary>
+        public static void Enable()
+        {
+            if (Interlocked.CompareExchange(ref s_enabled, 1, 0) == 0)
+                AppDomain.CurrentDomain.AssemblyResolve += CurrentDomain_AssemblyResolve;
+        }
+
+        private static Assembly CurrentDomain_AssemblyResolve(object sender, ResolveEventArgs args)
+        {
+            // apply any existing policy
+            AssemblyName referenceName = new AssemblyName(AppDomain.CurrentDomain.ApplyPolicy(args.Name));
+
+            // unify up to a local file if it exists.
+            string fileName = Path.Combine(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location), referenceName.Name + ".dll");
+            if (File.Exists(fileName))
+            {
+                var localName = AssemblyName.GetAssemblyName(fileName);
+
+                // if the local file has a version greater than what was
+                // requested assume it's backward compatible and load it
+                if (localName.Version > referenceName.Version)
```

Comment:
That is indeed what CoreCLR and the .NET loader in store apps does.  We cannot make .NET desktop do it for compat reasons.  As a result we have bindingRedirects on classic .NET but MSBuild doesn't expose bindingRedirects to task DLLs, so we have this workaround that mimics the CoreCLR behavior.